keps/sig-node/3066-subsecond-probes/README.md

```diff
@@ -0,0 +1,859 @@
+	}
+```
+
+ - Period Seconds is the biggest hurdle in charging and overall cost to the cluster, but also the most useful if set correctly.
+PeriodSeconds defaults to 10 seconds if unset (which includes the explicitly set to 0 state). Fast failure detections are thus blocked with existing defaults.
+https://github.com/kubernetes/kubernetes/blob/e19964183377d0ec2052d1f1fa930c4d7575bd50/pkg/apis/core/v1/defaults.go#L227-L229
+
+```go
+	if obj.PeriodSeconds == 0 {
+		obj.PeriodSeconds = 10
+	}
+```
+
+#### Validation of fields
+
+Must be non-negative, Zero or greater.
+```go
+	allErrs = append(allErrs, ValidateNonnegativeField(int64(probe.InitialDelaySeconds), fldPath.Child("initialDelaySeconds"))...)
+	allErrs = append(allErrs, ValidateNonnegativeField(int64(probe.TimeoutSeconds), fldPath.Child("timeoutSeconds"))...)
+	allErrs = append(allErrs, ValidateNonnegativeField(int64(probe.PeriodSeconds), fldPath.Child("periodSeconds"))...)
+
+```
+
+#### Fields to Add
+What fields may be necessary to add?
+
+In pkg/apis/core/types.go
+
+```go
+
+type Probe struct {
+    ...
+    ...
+    // The offset in milliseconds to PeriodSeconds to change how often to perform the probe before it succeeded.
+	// +optional
+	PeriodMilliseconds *int32
+	// Length of time (in milliseconds) to offset IntialDelaySeconds before health checking is activated.
+	// +optional
+	InitialDelayMilliseconds *int32
+}
+
+```
+#### Logic for Added Fields
+What is the least logic that could be used?
+
+The combined value of `PeriodSeconds` and `PeriodMilliseconds` must be greater than 200ms and greater than 500ms for exec probe. And after success falls back to original seconds value.
+The combined value of `IntialDelaySeconds` and `InitialDelayMilliseconds` must be greater than or equal to zero ms.
+
+
+### Existing use of Probe struct fields.
+
+#### InitialDelaySeconds
+https://github.com/kubernetes/kubernetes/blob/7c2e6125694e1aadc78a5fed1cf696872af50a5e/pkg/kubelet/prober/worker.go#L246-L249
+```go
+	// Probe disabled for InitialDelaySeconds.
+	if int32(time.Since(c.State.Running.StartedAt.Time).Seconds()) < w.spec.InitialDelaySeconds {
+		return true
+	}
+```
+
+#### TimeoutSeconds
+https://github.com/kubernetes/kubernetes/blob/7c2e6125694e1aadc78a5fed1cf696872af50a5e/pkg/kubelet/prober/prober.go#L160-L213
+```go
+	timeout := time.Duration(p.TimeoutSeconds) * time.Second
+```
+
+#### ProbeSeconds
+https://github.com/kubernetes/kubernetes/blob/7c2e6125694e1aadc78a5fed1cf696872af50a5e/pkg/kubelet/prober/worker.go#L131-L167
+```go
+func (w *worker) run() {
+	probeTickerPeriod := time.Duration(w.spec.PeriodSeconds) * time.Second
+
+	// If kubelet restarted the probes could be started in rapid succession.
+	// Let the worker wait for a random portion of tickerPeriod before probing.
+	// Do it only if the kubelet has started recently.
+	if probeTickerPeriod > time.Since(w.probeManager.start) {
+		time.Sleep(time.Duration(rand.Float64() * float64(probeTickerPeriod)))
+	}
```

Comment:
which period does this use for jitter? The fast initial one, or the steady state?

Reply:
period is after initial.. thus jitter is being applied to steady state..